### ISSUE TITLE
backward pass of `ch.nn.functional.scaled_dot_product_attention` is not deterministic for `bfloat16` / `float16` inputs

### ISSUE THREAD

**Issue by lengstrom** [pytorch/pytorch#123656]
### 🐛 Describe the bug

The backward pass of `ch.nn.functional.scaled_dot_product_attention` is not deterministic for `bfloat16` and `float16` inputs. The function is deterministic for `float32` backward passes, however. In particular the gradient wrt the `query` input is not deterministic (maybe the others aren't too but I did not find any inputs for which they weren't). Below is a minimal example.

```python
import torch as ch

def set_seed(seed):
    ch.manual_seed(seed)
    ch.cuda.manual_seed(seed)
    ch.cuda.manual_seed_all(seed)
    ch.backends.cudnn.deterministic = True
    ch.backends.cudnn.benchmark = False
    ch.backends.cuda.matmul.allow_tf32 = True
    ch.backends.cudnn.allow_tf32 = True
    ch.backends.cudnn.enabled = True
    ch.use_deterministic_algorithms(True)

def get_quantities(q, k, v, dtype):
    with ch.cuda.amp.autocast(dtype=dtype):
        y = ch.nn.functional.scaled_dot_product_attention(q, k, v,
                                                            attn_mask=None,
                                                            dropout_p=0,
                                                            is_causal=True) 
    ys = y.sum()
    ys.sum().backward()

    outputs = [y, q.grad, k.grad, v.grad]
    outputs = [v.clone().detach().cpu() for v in outputs]
    return list(zip(['y', 'grad q', 'grad k', 'grad v'], outputs))

def rand_tensor():
    v = ch.randn(32, 16, 511, 64, device='cuda')
    v.requires_grad_(True)
    return v

def main(dtype):
    for seed in range(3):
        set_seed(seed)
        q, k, v = rand_tensor(), rand_tensor(), rand_tensor()
        outputs1 = get_quantities(q, k, v, dtype)

        set_seed(seed)
        q, k, v = rand_tensor(), rand_tensor(), rand_tensor()
        
        outputs2 = get_quantities(q, k, v, dtype)
        failed = False
        for (n1, o1), (_, o2) in zip(outputs1, outputs2):
            max_diff = (o1-o2).absolute().max().item()
            if max_diff != 0:
                print(f'>> {dtype}: failed for seed', seed, 'with max diff',
                      max_diff, 'at tensor', n1)
                failed = True

    if not failed:
        print(f'>> {dtype}: all tests passed!')


main(ch.bfloat16)
main(ch.float32)
```
yields
```
>> torch.bfloat16: failed for seed 0 with max diff 0.0009765625 at tensor grad q
>> torch.bfloat16: failed for seed 1 with max diff 0.0009765625 at tensor grad q
>> torch.bfloat16: failed for seed 2 with max diff 0.00048828125 at tensor grad q
>> torch.float32: all tests passed!
>> torch.float16: failed for seed 0 with max diff 0.0001220703125 at tensor grad q
>> torch.float16: failed for seed 1 with max diff 0.0001220703125 at tensor grad q
>> torch.float16: failed for seed 2 with max diff 0.000244140625 at tensor grad q
```

### Versions

```
Collecting environment information...
PyTorch version: 2.2.1+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.26.3
Libc version: glibc-2.35

Python version: 3.10.13 (main, Sep 11 2023, 13:44:35) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-100-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.1.105
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA A100 80GB PCIe
GPU 1: NVIDIA A100 80GB PCIe
GPU 2: NVIDIA A100 80GB PCIe
GPU 3: NVIDIA A100 80GB PCIe
GPU 4: NVIDIA A100 80GB PCIe
GPU 5: NVIDIA A100 80GB PCIe
GPU 6: NVIDIA A100 80GB PCIe
GPU 7: NVIDIA A100 80GB PCIe

Nvidia driver version: 535.54.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      48 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             128
On-line CPU(s) list:                0-127
Vendor ID:                          AuthenticAMD
Model name:                         AMD EPYC 7513 32-Core Processor
CPU family:                         25
Model:                              1
Thread(s) per core:                 2
Core(s) per socket:                 32
Socket(s):                          2
Stepping:                           1
Frequency boost:                    enabled
CPU max MHz:                        3681.6399
CPU min MHz:                        1500.0000
BogoMIPS:                           5190.19
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 invpcid_single hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca
Virtualization:                     AMD-V
L1d cache:                          2 MiB (64 instances)
L1i cache:                          2 MiB (64 instances)
L2 cache:                           32 MiB (64 instances)
L3 cache:                           256 MiB (8 instances)
NUMA node(s):                       2
NUMA node0 CPU(s):                  0-31,64-95
NUMA node1 CPU(s):                  32-63,96-127
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Mitigation; safe RET
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] onnx==1.14.0
[pip3] onnxruntime==1.15.1
[pip3] pytorch-pfn-extras==0.7.6
[pip3] pytorch-ranger==0.1.1
[pip3] torch==2.2.1
[pip3] torch-optimizer==0.3.0
[pip3] torchmetrics==1.3.1
[pip3] torchvision==0.17.1
[pip3] triton==2.2.0
[pip3] triton-pre-mlir==2.0.0
[conda] cudatoolkit               10.1.243             h6bb024c_0
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] pytorch-pfn-extras        0.7.6                    pypi_0    pypi
[conda] pytorch-ranger            0.1.1                    pypi_0    pypi
[conda] torch                     2.2.1                    pypi_0    pypi
[conda] torch-optimizer           0.3.0                    pypi_0    pypi
[conda] torchmetrics              1.3.1                    pypi_0    pypi
[conda] torchvision               0.17.1                   pypi_0    pypi
[conda] triton                    2.2.0                    pypi_0    pypi
[conda] triton-pre-mlir           2.0.0                    pypi_0    pypi
```

cc @ptrblck @mruberry @kurtamohler @jbschlosser @bhosmer @cpuhrsch @erichan1 @drisspg @mikaylagawarecki

**Comment by lengstrom** [comment]
I think this is just a problem with the flash attention kernel, when I disable it all the tests pass.

**Comment by drisspg** [comment]
Could you try this on nightly? I am unable to reproduce, since nightly has this change:
https://github.com/pytorch/pytorch/blob/56dd7603da0d9f8b7f7439cf2091a1fd306983f5/aten/src/ATen/native/transformers/cuda/attention_backward.cu#L80-L90

**Comment by lengstrom** [comment]
on nightly it succeeds, thank you!
